Bubble up errors in event handler.

wen-coding opened the pull request.

AshwinSekar reviewed (approved):
Thanks for cleaning this up,
just as a broader sanity check is there any reason that `insert_vote_and_create_bls_message` can fail for some sort of transient error? or any error from this function indicates that we something is seriously wrong and we should shutdown.

AshwinSekar reviewed (changes requested):
Actually sorry approved too early, if i'm tracing the code correctly this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?

Since we're here do you mind fixing it? Also doesn't have to be in this PR but would be nice to have an error.rs to move all the various Error types to

wen-coding commented:
> Actually sorry approved too early, if i'm tracing the code correctly this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?
> 
> Since we're here do you mind fixing it? Also doesn't have to be in this PR but would be nice to have an error.rs to move all the various Error types to

That's a good point. I assume any error during generate_vote_tx is transient? I've changed them to an error log.

AshwinSekar commented:
> > Actually sorry approved too early, if i'm tracing the code correctly this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?
> > Since we're here do you mind fixing it? Also doesn't have to be in this PR but would be nice to have an error.rs to move all the various Error types to
> 
> That's a good point. I assume any error during generate_vote_tx is transient? I've changed them to an error log.

So `NonVoting` and `HotSpare` are normal for RPC / hot spare setups `info!` is fine here. 
`Failed` is due to startup scenarios (verification or wait for slot), or situations where the vote account is not setup / other misconfigurations. I would use `warn! `here, or if we really wanted to be precise split this into two enums and warn for the startup related ones and error for the misconfigurations.

Not sure where NoRankFound should fit in, I assume this can also happen if the validator does not meet the top 2k cap? Or even then would we expect the rank to exist?

wen-coding commented:
NoRankFound can happen if you are scheduled to vote for next Epich, so in
the current Epoch you can’t vote, so I’ll also mark it transient. Can’t ask
users to set up voting at the last minute of current Epoch.

On Sun, Aug 10, 2025 at 18:26 Ashwin Sekar ***@***.***> wrote:

> *AshwinSekar* left a comment (anza-xyz/alpenglow#386)
> <https://github.com/anza-xyz/alpenglow/pull/386#issuecomment-3173053944>
>
> Actually sorry approved too early, if i'm tracing the code correctly this
> still causes hot spares to shutdown right? Since any generate_vote_tx value
> that's not BLSMessage would result in a vote error?
> Since we're here do you mind fixing it? Also doesn't have to be in this PR
> but would be nice to have an error.rs to move all the various Error types
> to
>
> That's a good point. I assume any error during generate_vote_tx is
> transient? I've changed them to an error log.
>
> So NonVoting and HotSpare are normal for RPC / hot spare setups info! is
> fine here.
> Failed is due to startup scenarios (verification or wait for slot), or
> situations where the vote account is not setup / other misconfigurations. I
> would use warn! here, or if we really wanted to be precise split this
> into two enums and warn for the startup related ones and error for the
> misconfigurations.
>
> Not sure where NoRankFound should fit in, I assume this can also happen if
> the validator does not meet the top 2k cap? Or even then would we expect
> the rank to exist?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/anza-xyz/alpenglow/pull/386#issuecomment-3173053944>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A3FJ5FNXSN7KNAXFAMNO4CT3M7WNTAVCNFSM6AAAAACDPFROPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNZTGA2TGOJUGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

wen-coding commented:
> > > Actually sorry approved too early, if i'm tracing the code correctly this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?
> > > Since we're here do you mind fixing it? Also doesn't have to be in this PR but would be nice to have an error.rs to move all the various Error types to
> > 
> > 
> > That's a good point. I assume any error during generate_vote_tx is transient? I've changed them to an error log.
> 
> So `NonVoting` and `HotSpare` are normal for RPC / hot spare setups `info!` is fine here. `Failed` is due to startup scenarios (verification or wait for slot), or situations where the vote account is not setup / other misconfigurations. I would use `warn! `here, or if we really wanted to be precise split this into two enums and warn for the startup related ones and error for the misconfigurations.
> 
> Not sure where NoRankFound should fit in, I assume this can also happen if the validator does not meet the top 2k cap? Or even then would we expect the rank to exist?

Okay, I split GenerationError into GenerationInfo and GenerationWarning respectively.

bw-solana commented on votor/src/voting_utils.rs:
```diff
@@ -30,8 +30,10 @@ pub enum GenerateVoteTxResult {
+    WaitToVoteSlot,
     // failed generation, eligible for refresh
-    Failed,
+    Failed(String),
```
Should we call this `FailedToGenerate`? Because it looks like we call out some specific failures in the enum but then lump some into this `Failed` enum and then differentiate based on the `String`

bw-solana replied:
Related - would it make more sense to call out all the appropriate categories in the enum? Or possibly a sub-enum? And then we could attach the error strings to the enum, similar to what was done for `VoteError`

bw-solana commented on votor/src/voting_utils.rs:
```diff
@@ -63,11 +65,20 @@ pub enum BLSOp {
-    #[error("Unable to generate bls vote message")]
-    GenerationError(Box<GenerateVoteTxResult>),
+    #[error("Unable to generate bls vote message, transient error: {0:?}")]
+    GenerationInfo(Box<GenerateVoteTxResult>),
+
+    #[error("Unable to generate bls vote message, configuration error: {0:?}")]
+    GenerationWarning(Box<GenerateVoteTxResult>),
```
I think coupling these errors to the log level isn't the best.

I would propose naming `TransientGeneration` and `InvalidConfig`

wen-coding replied:
I assume you mean TransientError, changed.

bw-solana replied:
yeah, `TransientError` works

bw-solana commented on votor/src/event_handler.rs:
```diff
@@ -388,10 +414,10 @@ impl EventHandler {
         parent_block @ (parent_slot, parent_block_id): Block,
```
we should touch up the function comment here to indicate what the returned `Result` and `bool` mean

bw-solana commented on core/src/replay_stage.rs:
```diff
@@ -2770,21 +2770,26 @@ impl ReplayStage {
         if let Some(slot) = wait_to_vote_slot {
             if bank.slot() < slot {
-                return GenerateVoteTxResult::Failed;
+                return GenerateVoteTxResult::Failed(format!(
```
should this be `GenerateVoteTxResult::WaitToVoteSlot`?

wen-coding replied:
Oops, good catch, changed.

AshwinSekar reviewed (approved):
overall LGTM

carllin commented:
>  this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?

What is the danger here, is this fixing an existing issue?

wen-coding commented:
> > this still causes hot spares to shutdown right? Since any generate_vote_tx value that's not BLSMessage would result in a vote error?
> 
> What is the danger here, is this fixing an existing issue?

This is not fixing an existing issue, just change the event_handler.rs to follow normal Result<..., Error> handling, and distinguish between InvalidConfig and TransientError explicitly.

We want hot spares to not vote, do we want them to shut down? Then they are no longer "hot" spares.
The only behavior change is:
- startup verification incomplete will become an info log, not error log and not exiting (at Alpenglow side)
I don't think there is any behavior change on TowerBFT side.

carllin commented on votor/src/voting_utils.rs:
```diff
@@ -262,3 +275,28 @@ pub(crate) fn insert_vote_and_create_bls_message(
 }
+
+pub fn generate_vote_message_and_push_to_votes(
+    votes: &mut Vec<BLSOp>,
```
I think it would be cleaner to return `Result<Option<BLSOp>, VoteError> {` and have the caller run something like

```
if let Some(bls_op) = generate_vote_message(vote, is_refresh, vctx)? {
    votes.push(bls_op);
}
```

wen-coding replied:
Sure, done.

carllin commented:
> and not exiting (at Alpenglow side)

Ah because `WaitForStartupVerification` now maps to a `VoteError::TransientError` which is ignored and only logged?

This seems like fixing an existing bug, could we update the PR description to capture these

wen-coding commented:
> > and not exiting (at Alpenglow side)
> 
> Ah because `WaitForStartupVerification` now maps to a `VoteError::TransientError` which is ignored and only logged?
> 
> This seems like fixing an existing bug, could we update the PR description to capture these

Actually, I think we might currently exit on any transient error on Alpenglow side:
https://github.com/anza-xyz/alpenglow/blob/be34e2f7690f25c3d70c753f6c92ca174257ef2b/votor/src/voting_utils.rs#L236
This will return GenerationError when the generation result is not a BLSMessage

Then the Errors are pushed into votes:
https://github.com/anza-xyz/alpenglow/blob/be34e2f7690f25c3d70c753f6c92ca174257ef2b/votor/src/event_handler.rs#L256

Then we exit on any error inside votes:
https://github.com/anza-xyz/alpenglow/blob/be34e2f7690f25c3d70c753f6c92ca174257ef2b/votor/src/event_handler.rs#L155

And the other problem is we will not exit as soon as there is a non-transient error.

Updating PR description.

carllin commented on votor/src/voting_utils.rs:
```diff
@@ -262,3 +275,26 @@ pub(crate) fn insert_vote_and_create_bls_message(
     })
 }
+
+pub fn generate_vote_message_and_push_to_votes(
```
nit: function name can be generate_vote_message()

carllin commented on votor/src/voting_utils.rs:
```diff
@@ -233,8 +249,14 @@ pub(crate) fn insert_vote_and_create_bls_message(
     let bls_message = match generate_vote_tx(&vote, &bank, context) {
         GenerateVoteTxResult::BLSMessage(bls_message) => bls_message,
+        e @ GenerateVoteTxResult::NoAuthorizedVoter(_, _)
+        | e @ GenerateVoteTxResult::NoVoteState(_)
+        | e @ GenerateVoteTxResult::VoteAccountNotFound(_)
+        | e @ GenerateVoteTxResult::Tx(_) => {
+            return Err(VoteError::InvalidConfig(Box::new(e)));
+        }
         e => {
-            return Err(VoteError::GenerationError(Box::new(e)));
+            return Err(VoteError::TransientError(Box::new(e)));
```
so is essentially everything either an InvalidConfig or TransientError? so generate_vote_message_and_push_to_votes() will essentially never bubble up an error?

carllin commented on votor/src/voting_utils.rs:
```diff
@@ -233,8 +249,14 @@ pub(crate) fn insert_vote_and_create_bls_message(
+        | e @ GenerateVoteTxResult::VoteAccountNotFound(_)
+        | e @ GenerateVoteTxResult::Tx(_) => {
+            return Err(VoteError::InvalidConfig(Box::new(e)));
+        }
```
For maintainability/avoiding shooting ourselves in the foot, it would be nice to force anybody who adds a new `GenerateVoteTxResult` to bucket it into the right category

Maybe add a `is_invalid_config()`  and `is_transient_error` functions on `GenerateVoteTxResult` that goes through every case?

wen-coding replied:
Done.
